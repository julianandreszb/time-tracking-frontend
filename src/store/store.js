import Vue from 'vue'
import Vuex from 'vuex'
import appModule from "@/store/modules/appModule";

Vue.use(Vuex)

export default new Vuex.Store({
    modules: {
        appModule
    },
})
