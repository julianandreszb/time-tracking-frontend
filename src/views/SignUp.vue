<template>
  <v-app id="inspire">
    <v-main>
      <v-container fluid fill-height>
        <v-layout align-center justify-center>
          <v-flex xs12 sm8 md4>
            <v-card class="elevation-12">
              <v-toolbar dark color="primary">
                <v-toolbar-title>SIGN UP</v-toolbar-title>
              </v-toolbar>
              <v-card-text>
                <v-form
                    ref="form"
                    v-model="valid"
                    lazy-validation
                >
                  <v-text-field
                      prepend-icon="mdi-account"
                      id="v-text-field-name"
                      v-model="name"
                      :counter="20"
                      :rules="nameRules"
                      label="Name"
                      required
                  />
                  <v-text-field
                      prepend-icon="mdi-email"
                      id="v-text-field-email"
                      v-model="email"
                      :rules="emailRules"
                      label="E-mail"
                      required
                  />
                  <v-text-field
                      prepend-icon="mdi-lock"
                      id="v-text-field-password"
                      v-model="password"
                      :append-icon="showPassword ? 'mdi-eye' : 'mdi-eye-off'"
                      :rules="passwordRules"
                      :type="showPassword ? 'text' : 'password'"
                      label="Password"
                      counter
                      @click:append="showPassword = !showPassword"
                  ></v-text-field>
                  <v-text-field
                      prepend-icon="mdi-lock"
                      id="v-text-field-confirm-password"
                      v-model="passwordConfirmation"
                      :append-icon="showPassword ? 'mdi-eye' : 'mdi-eye-off'"
                      :rules="[passwordConfirmationRules.required, passwordConfirmationRule]"
                      :type="showPassword ? 'text' : 'password'"
                      label="Confirm Password"
                      counter
                      @click:append="showPassword = !showPassword"
                  ></v-text-field>
                </v-form>
              </v-card-text>
              <v-card-actions>
                <v-btn id="v-btn-signup" block color="primary" @click="submitSignUpForm">SIGN UP</v-btn>
              </v-card-actions>
              <v-card-actions>
                <v-spacer></v-spacer>
                <p class="my-2">Have already an account?</p>
                <v-btn
                    class="ma-1"
                    color="primary"
                    plain
                    to="/signin"
                >
                  Login here
                </v-btn>
                <v-spacer></v-spacer>
              </v-card-actions>
            </v-card>
          </v-flex>
          <DialogLoading
              :isDisplayed="displayDialogLoading"
              text="Creating new user"
          />
          <DialogError
              :title="dialogErrorTitle"
              :text="dialogErrorText"
              :isDisplayed="displayDialogError"

              @close="hideDialogError"
              data-testid="dialog-error"
          >
            <ErrorListHelper :dialogErrorJson="dialogErrorJson"/>
          </DialogError>
          <DialogInformation
              :title="dialogInformationTitle"
              :text="dialogInformationText"
              :isDisplayed="displayDialogInformation"
              @close="handleConfirmUserCreated"
          />
        </v-layout>
      </v-container>
    </v-main>
  </v-app>
</template>

<script>
import {mapGetters} from 'vuex'
import {signUpRequest} from "@/requests/signUpRequests";
import {signUpData} from '@/Shared/auth/authData'

import DialogLoading from "@/components/DialogLoading/DialogLoading";
import dialogLoadingData from "@/components/DialogLoading/dialogLoadingData";
import {showDialogLoading, hideDialogLoading} from '@/components/DialogLoading/dialogLoadingMethods'

import DialogError from "@/components/DialogError/DialogError";
import dialogErrorData from "@/components/DialogError/dialogErrorData";
import {hideDialogError, showDialogError} from '@/components/DialogError/dialogErrorMethods'

import DialogInformation from "@/components/DialogInformation/DialogInformation";
import dialogInformationData from "@/components/DialogInformation/dialogInformationData";
import {showDialogInformation} from '@/components/DialogInformation/dialogInformationMethods'

import ErrorListHelper from "@/components/ErrorListHelper/ErrorListHelper";
import errorListHelperData from "@/components/ErrorListHelper/errorListHelperData";

import {
  handleSignUpErrorResponse,
  handleSignUpSuccessResponse
} from '@/Shared/auth/authMethods'

export default {
  name: "SignUp",
  data: () => ({
    ...dialogErrorData,
    ...dialogInformationData,
    ...dialogLoadingData,
    ...errorListHelperData,
    ...signUpData
  }),
  components: {
    ErrorListHelper,
    DialogInformation,
    DialogLoading,
    DialogError
  },
  methods: {
    hideDialogError,
    showDialogError,

    showDialogInformation,

    showDialogLoading,
    hideDialogLoading,

    handleSignUpErrorResponse,
    handleSignUpSuccessResponse,

    handleConfirmUserCreated() {
      this.closeDialogInformation();

      this.$router.push('/signin');
    },
    submitSignUpForm() {
      const signUpFormIsValid = this.$refs.form.validate();
      this.isDebugEnabled && console.log('signUpFormIsValid', signUpFormIsValid);

      if (signUpFormIsValid) {

        this.showDialogLoading();

        signUpRequest({
          name: this.name,
          email: this.email,
          password: this.password,
          passwordConfirmation: this.passwordConfirmation
        }).catch(
            this.handleSignUpErrorResponse
        ).then(
            this.handleSignUpSuccessResponse
        );

      }
    },

  },
  computed: {
    ...mapGetters("appModule", [
      "isDebugEnabled"
    ]),
    passwordConfirmationRule() {
      return this.password === this.passwordConfirmation || 'Password must match';
    },
  }
}
</script>
