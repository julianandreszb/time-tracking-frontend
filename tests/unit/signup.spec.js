import {createLocalVue, RouterLinkStub, mount} from '@vue/test-utils'
import SignUp from "@/views/SignUp";
import Vuetify from "vuetify";
import store from "@/store/store";

describe('SignUp', () => {

    const localVue = createLocalVue();
    let vuetify
    let wrapper;

    beforeEach(() => {
        vuetify = new Vuetify()
        wrapper = mount(SignUp, {
            localVue,
            vuetify,
            store,
            stubs: {
                RouterLink: RouterLinkStub
            }
        });
    });

    test('snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('default signup form field values', () => {
        expect(wrapper.vm.name).toBe('');
        expect(wrapper.vm.email).toBe('');
        expect(wrapper.vm.password).toBe('');
        expect(wrapper.vm.passwordConfirmation).toBe('');
    });

    test('required field messages', async () => {

        wrapper.setData({
            name: '',
            email: '',
            password: '',
            passwordConfirmation: ''
        });

        await wrapper.find('#v-btn-signup').trigger('click');
        const signUpFormHtml = signUpFormContent();

        expect(signUpFormHtml).toContain('Name is required');
        expect(signUpFormHtml).toContain('E-mail is required');
        expect(signUpFormHtml).toContain('Password is required');
        expect(signUpFormHtml).toContain('Password Confirmation is required');

    });

    test('fill valid values into form input fields', () => {

        const wrapperName = wrapper.find('#v-text-field-name');
        const wrapperEmail = wrapper.find('#v-text-field-email');
        const wrapperPassword = wrapper.find('#v-text-field-password');
        const wrapperPasswordConfirm = wrapper.find('#v-text-field-confirm-password');

        wrapperName.element.value = 'Neil DeGrasse Tyson';
        wrapperEmail.element.value = 'neil_degrasse_tyson@test.com';
        wrapperPassword.element.value = 'ABC123456789';
        wrapperPasswordConfirm.element.value = 'ABC123456789';

        wrapperName.trigger('input');
        wrapperEmail.trigger('input');
        wrapperPassword.trigger('input');
        wrapperPasswordConfirm.trigger('input');

        wrapper.find('#v-btn-signup').trigger('click'); // wrapper.vm.$refs.form.validate();

        expect(wrapper.vm.name).toBe('Neil DeGrasse Tyson');
        expect(wrapper.vm.email).toBe('neil_degrasse_tyson@test.com');
        expect(wrapper.vm.password).toBe('ABC123456789');
        expect(wrapper.vm.passwordConfirmation).toBe('ABC123456789');

        const signUpFormHtml = signUpFormContent();
        expect(signUpFormHtml).not.toContain('Name is required');
        expect(signUpFormHtml).not.toContain('E-mail is required');
        expect(signUpFormHtml).not.toContain('Password is required');
        expect(signUpFormHtml).not.toContain('Password Confirmation is required');
        expect(signUpFormHtml).not.toContain('Password must match');

    });

    function signUpFormContent() {
        return wrapper.getComponent({ref: 'form'}).html();
    }

});
