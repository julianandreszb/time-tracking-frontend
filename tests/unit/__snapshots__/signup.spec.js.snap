// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp snapshot 1`] = `
<div data-app="true" class="v-application v-application--is-ltr theme--light" id="inspire">
  <div class="v-application--wrap">
    <main class="v-main" style="padding: 0px 0px 0px 0px;">
      <div class="v-main__wrap">
        <div class="container container--fluid fill-height">
          <div class="layout align-center justify-center">
            <div class="flex xs12 sm8 md4">
              <div class="elevation-12 v-card v-sheet theme--light">
                <header class="v-sheet theme--dark v-toolbar primary" style="height: 64px;">
                  <div class="v-toolbar__content" style="height: 64px;">
                    <div class="v-toolbar__title">SIGN UP</div>
                  </div>
                </header>
                <div class="v-card__text">
                  <form novalidate="novalidate" class="v-form">
                    <div class="v-input theme--light v-text-field">
                      <div class="v-input__prepend-outer">
                        <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-account theme--light"></i></div>
                      </div>
                      <div class="v-input__control">
                        <div class="v-input__slot">
                          <div class="v-text-field__slot"><label for="v-text-field-name" class="v-label theme--light" style="left: 0px; position: absolute;">Name</label><input required="required" id="v-text-field-name" type="text"></div>
                        </div>
                        <div class="v-text-field__details">
                          <div class="v-messages theme--light">
                            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                          </div>
                          <div class="v-counter theme--light">0 / 20</div>
                        </div>
                      </div>
                    </div>
                    <div class="v-input theme--light v-text-field">
                      <div class="v-input__prepend-outer">
                        <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-email theme--light"></i></div>
                      </div>
                      <div class="v-input__control">
                        <div class="v-input__slot">
                          <div class="v-text-field__slot"><label for="v-text-field-email" class="v-label theme--light" style="left: 0px; position: absolute;">E-mail</label><input required="required" id="v-text-field-email" type="text"></div>
                        </div>
                        <div class="v-text-field__details">
                          <div class="v-messages theme--light">
                            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div class="v-input theme--light v-text-field">
                      <div class="v-input__prepend-outer">
                        <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-lock theme--light"></i></div>
                      </div>
                      <div class="v-input__control">
                        <div class="v-input__slot">
                          <div class="v-text-field__slot"><label for="v-text-field-password" class="v-label theme--light" style="left: 0px; position: absolute;">Password</label><input id="v-text-field-password" type="password"></div>
                          <div class="v-input__append-inner">
                            <div class="v-input__icon v-input__icon--append"><button type="button" aria-label="append icon" class="v-icon notranslate v-icon--link mdi mdi-eye-off theme--light"></button></div>
                          </div>
                        </div>
                        <div class="v-text-field__details">
                          <div class="v-messages theme--light">
                            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                          </div>
                          <div class="v-counter theme--light">0</div>
                        </div>
                      </div>
                    </div>
                    <div class="v-input theme--light v-text-field">
                      <div class="v-input__prepend-outer">
                        <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-lock theme--light"></i></div>
                      </div>
                      <div class="v-input__control">
                        <div class="v-input__slot">
                          <div class="v-text-field__slot"><label for="v-text-field-confirm-password" class="v-label theme--light" style="left: 0px; position: absolute;">Confirm Password</label><input id="v-text-field-confirm-password" type="password"></div>
                          <div class="v-input__append-inner">
                            <div class="v-input__icon v-input__icon--append"><button type="button" aria-label="append icon" class="v-icon notranslate v-icon--link mdi mdi-eye-off theme--light"></button></div>
                          </div>
                        </div>
                        <div class="v-text-field__details">
                          <div class="v-messages theme--light">
                            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                          </div>
                          <div class="v-counter theme--light">0</div>
                        </div>
                      </div>
                    </div>
                  </form>
                </div>
                <div class="v-card__actions"><button type="button" class="v-btn v-btn--block v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary" id="v-btn-signup"><span class="v-btn__content">SIGN UP</span></button></div>
                <div class="v-card__actions">
                  <div class="spacer"></div>
                  <p class="my-2">Have already an account?</p> <a class="ma-1 v-btn v-btn--plain v-btn--router theme--light v-size--default primary--text"><span class="v-btn__content">
                Login here
              </span></a>
                  <div class="spacer"></div>
                </div>
              </div>
            </div>
            <div role="dialog" class="v-dialog__container">
              <!---->
            </div>
            <div role="dialog" class="v-dialog__container">
              <!---->
            </div>
            <div role="dialog" class="v-dialog__container">
              <!---->
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;
